src/Corax/Indexing/IndexWriter.FieldBuffers.cs

```diff
@@ -0,0 +1,138 @@
+﻿using System;
+using System.Buffers;
+using System.IO;
+using System.Runtime.InteropServices;
+using Voron;
+using Voron.Data.CompactTrees;
+using Voron.Data.Lookups;
+
+namespace Corax.Indexing;
+
+public partial class IndexWriter
+{
+    /// <summary>
+    /// FieldBuffers are used to prepare field terms in sorted order without allocating native memory and do not changing the orders of IndexedField properties since we're linking them via positions in buffers.
+    /// 
+    /// </summary>
+    private class FieldBuffers<TKey, TLookupKey> : IDisposable
+    where TKey : unmanaged
+    where TLookupKey : struct, ILookupKey
+    {
+        private readonly IndexWriter _parent;
+        public const int BatchSize = 1024;
+
+        private TKey[] _sortedTerms;
+        private int[] _termIndexes;
+
+        public TLookupKey[] Keys;
+        public int[] PageOffsets;
+        public long[] PostListIds;
+        private int[] _entriesOffsets;
+
+        public void PrepareTerms(IndexedField field, out Span<TKey> terms, out Span<int> indexes)
+        {
+            int termsCount;
+            if (typeof(TKey) == typeof(Slice))
+                termsCount = field.Textual.Count;
+            else if (typeof(TKey) == typeof(long))
+                termsCount = field.Longs.Count;
+            else if (typeof(TKey) == typeof(double))
+                termsCount = field.Doubles.Count;
+            else
+                throw new InvalidDataException($"Type {typeof(TKey).FullName} is not supported");
+            
+            if (_sortedTerms == null || _sortedTerms.Length < termsCount)
+            {
+                if (_sortedTerms != null)
+                {
+                    ArrayPool<TKey>.Shared.Return(_sortedTerms);
+                    ArrayPool<int>.Shared.Return(_termIndexes);
+                }
+
+                _sortedTerms = ArrayPool<TKey>.Shared.Rent(termsCount);
+                _termIndexes = ArrayPool<int>.Shared.Rent(termsCount);
+            }
+
+            int idx = 0;
+            if (typeof(TKey) == typeof(Slice))
+            {
+                foreach (var (k, v) in field.Textual)
+                {
+                    _sortedTerms[idx] = (TKey)(object)k;
+                    _termIndexes[idx] = v;
+                    idx++;
+                }
+            }
+            if (typeof(TKey) == typeof(long))
+            {
+                foreach (var (k, v) in field.Longs)
+                {
+                    _sortedTerms[idx] = (TKey)(object)k;
+                    _termIndexes[idx] = v;
+                    idx++;
+                }
+            }
+            if (typeof(TKey) == typeof(double))
+            {
+                foreach (var (k, v) in field.Doubles)
+                {
+                    _sortedTerms[idx] = (TKey)(object)k;
+                    _termIndexes[idx] = v;
+                    idx++;
+                }
+            }
+
```

Comment:
I would be interested to know how impactful is in term of performance the `PrepareTerms` call, because I see that `field.Longs` and `field.Doubles` can be improve with vector operations.

Reply:
How would you improve this with SIMD?